src/node/lattice.rs

```diff
@@ -298,30 +290,32 @@ impl ExecutionLattice {
             // of the earlier run_queue, clears the run_queue and initializes it afresh with the set
             // difference of the old run_queue and the nodes to remove.
             // Since the invocation of this code is hopefully rare, we can optimize it later.
-            if remove_roots.len() > 0 {
+            if demoted_roots.len() > 0 {
+                roots.retain(|e| !demoted_roots.contains(e));
                 // The run queue needs to be reconstructed.
-                let mut old_run_queue: Vec<RunnableEvent> = Vec::new();
-                for event in run_queue.iter() {
-                    old_run_queue.push(event.clone());
-                }
-                run_queue.clear();
-
-                roots.retain(|e| !remove_roots.contains(e));
-
+                let old_run_queue: Vec<RunnableEvent> = run_queue.drain().collect();
                 for item in old_run_queue {
-                    if !remove_roots.contains(&item) {
+                    if !demoted_roots.contains(&item) {
                         run_queue.push(item);
                     }
                 }
             }
 
-            // If no dependencies of this event, then we can safely create a new root in the forest and
+            // If the added event depends on no others, then we can safely create a new root in the forest and
             // add the event to the run queue.
-            if dependencies.is_empty() {
+            if preceding_events.is_empty() {
                 roots.push(RunnableEvent::new(event_ix).with_timestamp(event_timestamp.clone()));
                 run_queue.push(RunnableEvent::new(event_ix).with_timestamp(event_timestamp));
             }
         }
+
+        if forest.node_count() > 25 {
+            slog::warn!(
+                crate::get_terminal_logger(),
+                "{} operator events queued in lattice. Increase number of operator executors or decrease incoming \
```

Comment:
IMO, it would be useful to also include the operator name/id in this log message. We can do this in a different change because we'll also have to pass the id/name to the lattice.